pkg/fs/fs.go

```diff
@@ -15,3 +21,159 @@ func newFileExistsError(path string) FileExistsError {
 func (e FileExistsError) Error() string {
 	return fmt.Sprintf("operation not allowed, file %q exists", e.path)
 }
+
+type DiskStatus struct {
+	All   uint64
+	Used  uint64
+	Free  uint64
+	Avail uint64
+}
+
+// DirSize returns total size in bytes of containing files
+func DirSize(path string) (uint64, error) {
+	var size uint64
+	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			size += uint64(info.Size())
+		}
+		return err
+	})
+	return size, err
+}
+
+// CopyFile copies the contents of the file named src to the file named
+// by dst. The file will be created if it does not already exist. If the
+// destination file exists, all it's contents will be replaced by the contents
+// of the source file. The file mode will be copied from the source and
+// the copied data is synced/flushed to stable storage.
+func CopyFile(src, dst string) (err error) {
+	in, err := os.Open(src)
+	if err != nil {
+		return
+	}
+	defer in.Close()
+
+	out, err := os.Create(dst)
+	if err != nil {
+		return
+	}
+	defer func() {
+		if e := out.Close(); e != nil {
+			err = e
+		}
+	}()
+
+	_, err = io.Copy(out, in)
+	if err != nil {
+		return
+	}
+
+	err = out.Sync()
+	if err != nil {
+		return
+	}
+
+	si, err := os.Stat(src)
+	if err != nil {
+		return
+	}
+	err = os.Chmod(dst, si.Mode())
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// CopyDir recursively copies a directory tree, attempting to preserve permissions.
+// Source directory must exist, destination directory must *not* exist.
+// Symlinks are ignored and skipped.
+// dirRenameFunc allows renaming  directories. It must return orig name, if renaming is not done..
+// dirFilterFunc allows filtering out directories. Returned true means skipping the dir.
+// fileFilterFunc allows filtering out files. Returned true means skipping the file.
+func CopyDir(src string, dst string, dirRenameFunc func(path string) string, dirFilterFunc func(path string) bool, fileFilterFunc func(path string) bool) (err error) {
```

Comment:
Sure, mote tests are in our backlog, So far we focused on completing functionality with IT test with happy path and manual testing.

Reply:
On reflection, I see that the filtering functionality isn't going to be available in an off-the-shelf package. But then it's probably quite specific to the upgrade functionality, and would probably be better kept local to that functionality rather than exporting in a `pkg/fs` package that seems intended for general use.

Reply:
Ok, I will be moved..